### PR TITLE
Telechat: Assembly Upstreaming III

### PR DESCRIPTION
**Pull request by lukeg101**
This patch is the last major set of assembly changes for herd, In detail it adds:
- Inline barrel shifter support for LDRB/H and STRB/H instructions
- Added TBZ and TBNZ instructions to AArch64
- Added __(u)int128_t C type, no unittests (see commit message for details)
- Added STXP instruction to AArch64
- Added LDXP instruction to AArch64
- Added LDAXP and STLXP instructions to AArch64
- Updated LDAR semantics to handle numeric addresses in AArch64
- Updated STXR semantics to handle numeric addresses in AArch64
- Updated label support for BNE,BEQ,CBZ,CBNZ,B,and BL in AArch64
- Merged this branch with upstream fork to make pull request proceed without the need for a manual merge

Unit tests are provided unless otherwise specified.

There may be other minor assembly changes later down the line, but for now this covers most of them.

### CODE REVIEW

**Comment by lukeg101** [comment]
For clarity new commits are from `6e75bd9` onwards

File: herd/AArch64Sem.ml [hunk 1]
```diff
@@ -428,6 +428,39 @@ module Make
...
+          (fun ea1 resa1 v1 -> write sz ea1 v1 resa1 ii)
+          (* We increment the residual address here so we ask about the second element of the pair *)
+          (fun ea2 resa2 v2 ->
+            M.address_of resa2
+            >>= incr_a var
+            >>= M.deref
+            >>= fun resa2 -> write sz ea2 v2 resa2 ii)
+
+        >>! B.Next
+
```
[the remark this answers is not in the document]
Comment (maranget) [reply herd/AArch64Sem.ml:478]: Here we are, something is wrong; STXP tests the contents of  the reservation register twice (have a look at write_mem_atomic, which is called for addr1 and addr2. However, ldar writes a into the register. As a result we (may ?) get an equation a = incr a, which has no solution.
I'd suggest to store a in the reservation register, as (ldar now does) and to test the contents of the reservation register  against the base address only in stxp.

I also observe that no unit test tests a successful LDXP/STXP pair.

File: herd/unittests/AArch64/A29.litmus [hunk 2]
```diff
@@ -0,0 +1,10 @@
+AArch64 A29
+(* Test Store exclusive, concrete location *)
+
+ { 0:X0 = 1; 0:X1=2; z=0; &z=200; &a=264; 0:X2=z; 0:x3=2;}
+
+P0;
+  STXR W3, X0, [X2];
+
+(* We expect the instruction to fail without an accompanying load *)
+exists (z=0  /\ a=0 /\ 0:X3=1)
```
Comment (maranget) [inline herd/unittests/AArch64/A29.litmus:10]: Although not very important, the comment suggests to replace exists by forall in the condition.

File: herd/unittests/AArch64/A43.litmus [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+(* Test load exclusive; Store exclusive, symbolic location *)
+
+ { 0:X4=z; z=2}
+
+P0;
+  LDXR X0, [X4];
+  ADD X0, X0, #1;
+  STXR W3, X0, [X4];
+
+exists ((z=2 /\ 0:X3=1) \/ (z=3 /\ 0:X3=0))
```
Comment (maranget) [inline herd/unittests/AArch64/A43.litmus:11]: Again `forall` may be more appropriate.

File: gen/AArch64Compile_gen.ml [hunk 4]
```diff
@@ -248,8 +248,8 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
     let stlr_mixed_idx sz r1 r2 idx  =
```
[the remark this answers is not in the document]
Comment (lukeg101) [reply gen/AArch64Compile_gen.ml:261]: fair

File: herd/AArch64Sem.ml [hunk 5]
```diff
@@ -537,6 +537,19 @@ module Make
               >>= is_not_zero
               >>= fun v -> commit_bcc ii
                   >>= fun () -> B.bccT v l
+        | I_TBZ(_,r,k,l) ->
+            (read_reg_ord r ii)
+              >>= M.op Op.BitAt (V.intToV k)
```
[the remark this answers is not in the document]
Comment (lukeg101) [reply herd/AArch64Sem.ml:591]: fair, I will remove `BitAt` entirely and use `ReadBit` for `I_TBZ` and `I_TBNZ`

File: litmus/AArch64Compile_litmus.ml [hunk 6]
```diff
@@ -123,6 +123,17 @@ let pp_shifter = function
...
+    let tbz tr_lab memo v r k lbl =
+      let memo =
+        sprintf
+          (match v with
+          | V32 -> "%s ^wi0,#%d, %s"
+          | V64 -> "%s ^i0, #%d, %s")
+          memo k (A.Out.dump_label (tr_lab lbl)) in
+      { empty_ins with
+        memo; inputs=[r;]; outputs=[];
+        branch=[Next; Branch lbl] ; }
```
[the remark this answers is not in the document]
Comment (lukeg101) [reply litmus/AArch64Compile_litmus.ml:135]: added

File: lib/machSize.ml [hunk 7]
```diff
@@ -14,32 +14,36 @@
...
 
-type sz = Byte | Short | Word | Quad
+type sz = Byte | Short | Word | Quad | QuadWord
 
 let pp = function
   | Byte -> "byte"
   | Short -> "short"
   | Word -> "word"
   | Quad -> "quad"
+  | QuadWord -> "quadword"
```
[the remark this answers is not in the document]
Comment (lukeg101) [reply lib/machSize.ml:24]: I am going to revert the 128 bit patch and implement vectors, then we can use a Vec 2 Int64 to represent 128 bit types

File: herd/event.ml [hunk 8]
```diff
@@ -1251,6 +1261,98 @@ let do_para_comp es1 es2 =
...
+              then in_wresult else EventSet.empty));
+         EventRel.cartesian
+           (EventSet.union out_addr2 out_resa)
+           (EventSet.union3 in_wmem2 in_wres
+              (if C.variant Variant.FullScDepend || success
+              then in_wresult else EventSet.empty));
+         EventRel.cartesian out_data1
+           (if C.variant Variant.Success || not (C.variant Variant.FullScDepend) then
+              EventSet.union in_wmem1 in_wmem2 else
+           EventSet.union3 in_wresult in_wmem1 in_wmem2);
```
[the remark this answers is not in the document]
Comment (lukeg101) [reply herd/event.ml:1309]: This patch is also old, so I'm not sure either

File: herd/AArch64Sem.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukeg101) [reply herd/AArch64Sem.ml:478]: I am going to revert the patch for LDXP too until we have vectors - I won't address this feedback until I resubmit after vectors

File: herd/unittests/AArch64/A29.litmus (hunk 2 above)
Comment (lukeg101) [reply herd/unittests/AArch64/A29.litmus:10]: fixed

File: herd/unittests/AArch64/A43.litmus (hunk 3 above)
Comment (lukeg101) [reply herd/unittests/AArch64/A43.litmus:11]: fixed

File: herd/unittests/AArch64/A44.litmus [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+(* Test load exclusive; Store exclusive, symbolic location *)
+
+ { 0:X4=z; z=2; 1:X5=5; 1:X6=z}
+
+P0                    | P1;
+  LDXR X0, [X4]       | STR X5, [X6];
+  ADD X0, X0, #1      | ;
+  STXR W3, X0, [X4]   | ;
+
+exists ((not (z=3  /\ 0:X3=1)) \/ (not (z=6 /\ 0:X3=1)))
```
[the remark this answers is not in the document]
Comment (lukeg101) [reply herd/unittests/AArch64/A44.litmus:11]: fixed

**Comment by lukeg101** [comment]
I have addressed all non-`ld*p`/`st*p`/`__(u)int128_t` issues, and reverted all `st*p`/`ld*p`,`__(u)int128_t` patches. 

I will follow up this code review with vector support for size bounded vectors in herd, which we can then use as syntactic sugar for `__(u)int128_t` types, and handle ld/st pairs in a uniform fashion for herd and litmus

**Comment by lukeg101** [comment]
As such all controversial changes are removed from this patch, so unless there is any other issues I think this is ready to merge.

As a side note, I have kept ldp/stp/int128 as reverts in the commit history so that it is easy to get the patches later and adapt them to vectors

**Comment by maranget** [comment]
This is fine by me, let us merge.

**Pull request merged**